src/cascadia/WindowsTerminal/WindowEmperor.cpp

```diff
@@ -0,0 +1,730 @@
+        {
+            auto index = gsl::narrow_cast<int>(_hotkeys.size());
+            const auto succeeded = _registerHotKey(index, keyChord);
+
+            TraceLoggingWrite(g_hWindowsTerminalProvider,
+                              "AppHost_setupGlobalHotkey",
+                              TraceLoggingDescription("Emitted when setting a single hotkey"),
+                              TraceLoggingInt64(index, "index", "the index of the hotkey to add"),
+                              TraceLoggingWideString(cmd.Name().c_str(), "name", "the name of the command"),
+                              TraceLoggingBoolean(succeeded, "succeeded", "true if we succeeded"),
+                              TraceLoggingLevel(WINEVENT_LEVEL_VERBOSE),
+                              TraceLoggingKeyword(TIL_KEYWORD_TRACE));
+            _hotkeys.emplace_back(summonArgs);
+        }
+    }
+}
+
+#pragma endregion
+
+#pragma region NotificationIcon
+// Method Description:
+// - Creates a Notification Icon and hooks up its handlers
+// Arguments:
+// - <none>
+// Return Value:
+// - <none>
+void WindowEmperor::_createNotificationIcon()
+{
+    _notificationIcon = std::make_unique<NotificationIcon>(_window.get());
+    _notificationIcon->SummonWindowRequested([this](auto& args) { _manager.SummonWindow(args); });
+}
+
+// Method Description:
+// - Deletes our notification icon if we have one.
+// Arguments:
+// - <none>
+// Return Value:
+// - <none>
+void WindowEmperor::_destroyNotificationIcon()
+{
+    _notificationIcon->RemoveIconFromNotificationArea();
+    _notificationIcon = nullptr;
+}
+
+void WindowEmperor::_checkWindowsForNotificationIcon()
+{
+    // We need to check some conditions to show the notification icon.
+    //
+    // * If there's a Quake window somewhere, we'll want to keep the
+    //   notification icon.
+    // * There's two settings - MinimizeToNotificationArea and
+    //   AlwaysShowNotificationIcon. If either one of them are true, we want to
+    //   make sure there's a notification icon.
+    //
+    // If both are false, we want to remove our icon from the notification area.
+    // When we remove our icon from the notification area, we'll also want to
+    // re-summon any hidden windows, but right now we're not keeping track of
+    // who's hidden, so just summon them all. Tracking the work to do a "summon
+    // all minimized" in GH#10448
+    //
+    // To avoid races between us thinking the settings updated, and the windows
+    // themselves getting the new settings, only ask the app logic for the
+    // RequestsTrayIcon setting value, and combine that with the result of each
+    // window (which won't change during a settings reload).
+    bool needsIcon = _app.Logic().RequestsTrayIcon();
+    auto windows{ _windows.lock_shared() };
+    for (const auto& _windowThread : *windows)
+    {
+        needsIcon |= _windowThread->Logic().IsQuakeWindow();
+    }

```

Comment:
You can wrap this in `{}` to shorten the lock period and avoid the risk of renderer-style deadlocks.